### DISCUSSION TITLE
[java] Is there a way to say "type is X or there is no type information" on XPath Expression?

### DISCUSSION THREAD

**Discussion by contivero** [pmd/pmd#3294] [Q&A] [answered]
Hi. I've recently been writing a bunch of custom xpath rules, and wrote one which I wanted to apply to any Collection, so I used `pmd-java:typeIs("java.lang.Collection")`. This worked fine for basic cases, but when I tried it in a real codebase, I realized some cases weren't being flagged because the variable was being inherited from a parent, and PMD doesn't have any information about the superclass' fields (correct me if my assumption is wrong). Thus, I wanted a way of saying "if this has a type at all, it must be Collection; otherwise no type information is needed to match", but didn't find how I could express this. Is there a way? and if not, what do you think about adding one?

**Comment by oowekyala** [comment, marked as answer]
Currently, there's no way to do that within an XPath expression (though with java your hands are freer)

> PMD doesn't have any information about the superclass' fields (correct me if my assumption is wrong)

Actually it does, the type resolution framework does resolve them and use their type (but, it doesn't expose eg a reflect.Field instance, only the final Class for the expression). If the classpath is properly configured, and the type of the expression is not resolved, this is a bug that you should report to us :) In PMD 7 this part of pmd will also be very significantly improved.

Given that, I think it's a bad idea to add a way to do what you describe. If we have no information we could be reporting false positives. The correct way to fix these cases shouldn't be to make the rule less precise (which could backfire and you'd have a noisy rule that you won't pay attention to). But this hints that there may be a problem with our type resolution, or with your classpath configuration.

**Reply by contivero** [reply]
Thanks for the prompt reply. It might very well be a classpath issue on my side. I'm running PMD through one of the eclipse plugins, and it wouldn't be the first time that my classpath is messed up in Eclipse unfortunately. I'll play a bit more with it and write a proper bug report in the case I believe it's an issue on PMD's side.

> Given that, I think it's a bad idea to add a way to do what you describe.

I agree. Thanks for correcting my misunderstanding, and thank you for your work in such a great tool!
